### ISSUE TITLE
New sniff: alphabetize interfaces following `implements` keyword

### ISSUE THREAD

**Issue by ramsey** [PHPCSStandards/PHPCSExtra#54]
Similar to the [AlphabeticallySortedUsesSniff](https://github.com/slevomat/coding-standard/blob/master/SlevomatCodingStandard/Sniffs/Namespaces/AlphabeticallySortedUsesSniff.php) in slevomat/coding-standard, I'd like a sniff that checks whether interface names following the `implements` keyword are alphabetically sorted and, if not, fixes them, when running `phpcbf`.

For example, given the following class:

```php
class Baz implements Foo, Bar
{
}
```

This sniff would result in an error because `Foo` and `Bar` are not in alphabetical order.

Using `phpcbf` on this file would correct the code to:

```php
class Baz implements Bar, Foo
{
}
```

**Comment by jrfnl** [comment]
Hi Ben, I'd be happy to include such a sniff in PHPCSExtra.

Some questions which come to mind:

#### What about interfaces which extend multiple other interfaces ?
I think it would make sense to handle that cases too, i.e.:
```php
interface Foo extends Throwable, Countable {} // Incorrect order.
interface Foo extends Countable, Throwable {} // Correct order.
```

#### What should be the order when the `implements` or `extends` has both unqualified, fully qualified and/or partially qualified names ?

Base the order on the interface name (last part) or use "natural" alphabetic order (first part excluding leading backslash) ?
```php
class Baz implements Sub\Bar, Foo, \Vendor\Package\Boo {}
```

Should this be fixed to (interface name based):
```php
class Baz implements Sub\Bar, \Vendor\Package\Boo, Foo {}
```
or ("natural"):
```php
class Baz implements Foo, Sub\Bar, \Vendor\Package\Boo {}
```

#### How should the fixer handle comments and (additional) whitespace ?
Think a code sample like this:
```php
class Baz implements
    Sub\Bar, // Trailing comment.
    Foo,
    // Line comment.
    \Vendor\Package\Boo {}
```

It might be hard to determine which part of the statement the comment applies to, so should the fixer just leave this sample alone and let the sniff just throw an error ?


I'd be interested to hear your thoughts on these questions.

**Comment by jrfnl** [comment]
I've created a sniff for this, please see PR #55. Testing appreciated.

Regarding the questions I posed above - this is how I've implemented the sniff now:
* **What about interfaces which extend multiple other interfaces ?**
    I have included this in the sniff, but with a different error code, so the check can be easily in/excluded.
* **What should be the order when the implements or extends has both unqualified, fully qualified and/or partially qualified names ?**
    I have made the order configurable. Sorting can be either by the interface name alone (`name`) or by the full name as used in the statement (`full`).
    The default has been set to `name`.
* **How should the fixer handle comments and (additional) whitespace ?**
    If comments are found in the implements/extends statement, the sniff will not throw an error, but will not auto-fix.
    If additional whitespace is found (new lines etc) in the implements/extends statement, the sniff will remove it when auto-fixing. I.e. the fixer will separate each name with a comma and one space.
    If alternative formatting is desired, a sniff which will check and fix the formatting should be added to the ruleset.

**Comment by ramsey** [comment]
@jrfnl, Thank you for implementing this so quickly!

> **What about interfaces which extend multiple other interfaces?**
>
> I have included this in the sniff, but with a different error code, so the check can be easily in/excluded.

This sounds great!

> **What should be the order when the implements or extends has both unqualified, fully qualified and/or partially qualified names?**
>
> I have made the order configurable.

Perfect!

In practice, this is not something I do. I have other sniffs that disallow fully-qualified names (must include them with a `use` statement), so I had not considered this when requesting the feature.

> **How should the fixer handle comments and (additional) whitespace?**
>
> If comments are found in the implements/extends statement, the sniff will not throw an error, but will not auto-fix. 

This sounds like a good compromise. I also had not considered this use-case, since I don't use comments with `implements`.

**Comment by jrfnl** [comment]
> In practice, this is not something I do. I have other sniffs that disallow fully-qualified names (must include them with a use statement), so I had not considered this when requesting the feature.

> I also had not considered this use-case, since I don't use comments with implements.

I realize both of these will probably be uncommon situations, but a sniff should function independently of other sniffs, so can't presume too much, which is why I think the sniff should still take it into account.

I'll merge the sniff now and it will be included in the `1.0.0-alpha3` release which I'm hoping to tag soon.

Thanks for looking it over and for the sniff suggestion!
